src/llvm/analysis/ReachingDefinitions/ReachingDefinitions.cpp

```diff
@@ -273,92 +273,80 @@ RDNode *LLVMRDBuilder::createNode(const llvm::Instruction &Inst, RDBlock *rb)
+        printf("\t%lu: %p\n", i, inst->getOperand(i));
+    }
+}
 
-    pta::PSNode *pts = PTA->getPointsTo(Inst->getOperand(1));
-    assert(pts && "Don't have the points-to information for store");
-
-    if (pts->pointsTo.empty()) {
-        //llvm::errs() << "ERROR: empty STORE points-to: " << *Inst << "\n";
-
-        // this may happen on invalid reads and writes to memory,
-        // like when you try for example this:
-        //
-        //   int p, q;
-        //   memcpy(p, q, sizeof p);
-        //
-        // (there should be &p and &q)
-        // NOTE: maybe this is a bit strong to say unknown memory,
-        // but better be sound then incorrect
-        node->addDef(UNKNOWN_MEMORY);
-        return node;
+std::vector<DefSite> LLVMRDBuilder::getPointsTo(const llvm::Value *val, RDBlock *rb)
+{
+    std::vector<DefSite> result;
+
+    pta::PSNode *psn = PTA->getPointsTo(val);
+
+    if (!psn)
+        // don't have points-to information for used pointer
+        return result;
+
+    if (psn->pointsTo.empty()) {
+        result.push_back(DefSite(UNKNOWN_MEMORY));
     }
 
-    for (const pta::Pointer& ptr: pts->pointsTo) {
-        // XXX we should at least warn?
+    for (const pta::Pointer& ptr : psn->pointsTo) {
         if (ptr.isNull())
             continue;
 
         if (ptr.isUnknown()) {
-            node->addDef(UNKNOWN_MEMORY);
+            result.push_back(DefSite(UNKNOWN_MEMORY));
             continue;
         }
 
         const llvm::Value *ptrVal = ptr.target->getUserData<llvm::Value>();
-        // this may emerge with vararg function
         if (llvm::isa<llvm::Function>(ptrVal))
             continue;
 
         RDNode *ptrNode = getOperand(ptrVal, rb);
-        //assert(ptrNode && "Don't have created node for pointer's target");
         if (!ptrNode) {
-            // keeping such set is faster then printing it all to terminal
-            // ... and we don't flood the terminal that way
-            static std::set<const llvm::Value *> warned;
-            if (warned.insert(ptrVal).second) {
-                llvm::errs() << *ptrVal << "\n";
-                llvm::errs() << "Don't have created node for pointer's target\n";
-            }
-
+            // XXX: maybe warn here
+            llvm::errs() << *ptrVal << "\n";
+            llvm::errs() << "Don't have created node for pointer's source\n";
             continue;
         }
 
         uint64_t size;
         if (ptr.offset.isUnknown()) {
             size = UNKNOWN_OFFSET;
         } else {
-            size = getAllocatedSize(Inst->getOperand(0)->getType(), DL);
+            size = getAllocatedSize(val->getType(), DL);
             if (size == 0)
                 size = UNKNOWN_OFFSET;
         }
 
-        //llvm::errs() << *Inst << " DEFS >> " << ptr.target->getName() << " ["
-        //             << *ptr.offset << " - " << *ptr.offset + size - 1 << "\n";
-
-        // strong update is possible only with must aliases. Also we can not
```

Comment:
Maybe it would be valuable to keep such comments. I don't know, what do you think?